_posts/2020-07-12-the-lifecycle-of-spring-beans.md

```diff
@@ -0,0 +1,427 @@
+
+In Spring Framework, **hooking into the bean lifecycle is a good way to extend our application in most cases.**
+
+### Acquiring Bean Properties
+One of the use-cases is acquiring the bean properties (like bean name) in runtime. For example, when we do some logging:
+```java
+@Component
+public class NamedSpringBean implements BeanNameAware {
+
+  Logger logger = LoggerFactory.getLogger(NamedSpringBean.class);
+
+  public void setBeanName(String name) {
+    logger.info(name + " created.");
+  }
+
+}
+```
+### Dynamically Changing Spring Bean Instances
+In some cases, we need to define Spring beans programmatically. This can be a practical solution when we need to re-create and change our bean instances on-demand:
+```java
+@Service
+public class IpToLocationService implements BeanFactoryAware {
+
+  DefaultListableBeanFactory listableBeanFactory;
+  IpDatabaseRepository ipDatabaseRepository;
+
+  @Override
+  public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+    listableBeanFactory = (DefaultListableBeanFactory) beanFactory;
+    updateIpDatabase();
+  }
+
+  public void updateIpDatabase(){
+    String updateUrl = "https://download.acme.com/ip-database-latest.mdb";
+
+    AbstractBeanDefinition definition = BeanDefinitionBuilder
+        .genericBeanDefinition(IpDatabaseRepository.class)
+        .addPropertyValue("file", updateUrl)
+        .getBeanDefinition();
+
+    listableBeanFactory
+        .registerBeanDefinition("ipDatabaseRepository", definition);
+
+    ipDatabaseRepository = listableBeanFactory
+        .getBean(IpDatabaseRepository.class);
+  }
+}
+```
+### Accessing Beans From the Outside of the Spring Context
+Another scenario is accessing the `BeanFactory` instance from outside the Spring context.
+
+For example, we may want to inject `BeanFactory` to a non-spring class to be able to access Spring beans or configurations inside that class. Spring - Quartz integration is a good example to show this use-case:
+```java
+public class AutowireCapableJobFactory
+    extends SpringBeanJobFactory implements ApplicationContextAware {
+
+  private AutowireCapableBeanFactory beanFactory;
+
+  @Override
+  public void setApplicationContext(final ApplicationContext context) {
+    this.beanFactory = context.getAutowireCapableBeanFactory();
+  }
+
+  @Override
+  protected Object createJobInstance(final TriggerFiredBundle bundle)
+      throws Exception {
+    final Object job = super.createJobInstance(bundle);
+    this.beanFactory.autowireBean(job);
+    return job;
+  }
+
+}
+```
+Also, a common Spring - Jersey integration is another clear example of this:
+```java
+@Configuration
+public class JerseyConfig extends ResourceConfig {
+
+  @Autowired
+  private ApplicationContext applicationContext;
+
+  @PostConstruct
+  public void registerResources() {
+    applicationContext.getBeansWithAnnotation(Path.class).values()
+      .forEach(this::register);
+  }
+
+}
```

Comment:
Please add a sentence below each of the code snippets to explain what you're doing in the code